ocrd/ocrd/decorators/ocrd_cli_options.py

```diff
@@ -33,6 +34,8 @@ def cli(mets_url):
         parameter_option,
         parameter_override_option,
         loglevel_option,
+        option('--agent_type', type=click.STRING),
+        option('--agent_address', type=click.STRING),
```

Comment:
> So in short, **`this is still far away from a review, so please restrain to do so yet`**.

I hear you. The least thing I want is put extra work on your shoulders. The review was not meant that way though. I just tried to give some early commentary, in a mix of
- problems already visible, that will likely fall on your feet (e.g. the processor decorator/constructor contexts)
- small ideas you might find useful
- concerns (based on my limited understanding) which might indicate potential problems going forward

You don't need to address everything. But I'll keep quiet until you release the draft.

> The `client`, in the way you imagined it, is currently the `Processing Server` which redirects coming requests to deployed `Processor Servers` (i.e., processing worker as a standalone server - but both concepts are separated to avoid confusion).

I understand that. But the Processing Server **also** has value as a replacement for the processor CLI. It's complementary. IPs will have different models for different tasks – a full-blown Web-API backed Workflow Server with Processing Server and queues, a Processing Server without queues only Processor Servers, and finally just some scripts or custom tools with the simple CLI interface to processors. To make the latter at least more easily installable and scalable, a backwards compatible CLI interface to the Processor Servers makes sense, and would allow splitting up ocrd_all (now 36 GB!) into single module containers. The increasing problems getting ocrd_all workable in the current (fat container – few venvs) form is part of the urgency.

> See, even you have mistaken the usage between the `processor server` with the `processing worker`. The `processor server` does not take `--queue Q` as a parameter.

Yes, sorry, that was sloppy writing. Cut the `--queue Q` in my example.

> I think this is not something we would like to have in that form, do we? Correct me if I am wrong, but why would a user need to install the processor-executable on their machine only to be able to send a request to the `Processing Server` or `Processor Server` (which may not be running on their machine)? 

It wouldn't be an installation of the actual processor, just of core. Perhaps ocrd_all's recipes could install a client tool named after each processor to be installed, and the client tool looks up its name to determine where to connect. (So much for my idea.)

> How about just:
> 
> ```shell
> ocrd network client --address localhost:1000 -m path/to/mets.xml -I OCR-D-SEG -O OCR-D-OCR -P checkpoint_dir default
> ```

Yes, that would do it, too.

> Or:
> 
> ```shell
> ocrd network client --address localhost:1000 ocrd-calamari-recognize -m path/to/mets.xml -I OCR-D-SEG -O OCR-D-OCR -P checkpoint_dir default
> ```
> 
> Where `--address` is `Processing Server` address and the client doesn't have to bother knowing the exact address of the specific `Processor Server`. The `Processing Server` is a proxy in that case.

Yes, that's even better! So using the Processing Server as a proxy and deployer does have benefits even in this scenario.

> Of course, with proper env variables, we can also drop the `--address` option and resolve it internally.

Yes, I thought so, too. Something in the environment that will tell both the deployer and the client where everything can be found. Could also be the deployer config file, though – perhaps even for the client?

> In the end, these two will be equal:
> 
> ```shell
> ocrd network client ocrd-* -m path/to/mets.xml -I OCR-D-* -O OCR-D-* -P ...
> ocrd-* -m path/to/mets.xml -I OCR-D-* -O OCR-D-* -P ...
> ```

Yes, that's the idea. So our ocrd_all can just provide simple delegator scripts, and everything can be made to look like CLI installation, as before. (Of course, the module containers will have to be allowed to run as Processor Workers if required.)

> Isn't the `ocrd_cli_wrap_processor` method already cluttered enough when a processor can be a `worker` and a `server`, so we need a `client` on top with a different `--address` option as well? While having `worker` and `server` makes sense, since these cannot be called without having the executable in hand, that's not the case for the `client`.

If we do introduce a `network` command group, then – at least for the user – I don't think this is clutter. On the implementation side I don't see the problem either – the decorator just needs another conditional.

Reply:
Since it's not obvious how to implement `ocrd-* network ...` without changing separate processor implementations should we resolve this? I have already implemented the initial skeleton for the `ocrd network client ...` and the single processing request which seems to work. However, I think the discussion for that should continue under #1032. I will update the current CLI status there today or tomorrow.

Reply:
Yes, I agree we should resolve. Fighting for CLI grouping in our processor decorator is not worth the effort, if possible at all. And I think we already agreed that since we have a custom help generator which already can group the various options, true command grouping is not even necessary.